### PR TITLE
feat: add endpoint to return the `snapshotFee`

### LINKED ISSUE #95: feat: add nft-claimer api endpoint to return the `snapshotFee`
This value is not indexed by subgraph, and retrievable only by reading the smart contract.

Return it from the backend side, to avoid reading SC from the FE, as the FE may not be aware of  the SC address before asking for a mint/deploy payload

### PR DESCRIPTION
**Pull request by wa0x6e**
## 🧿 Current issues / What's wrong ?

- Space collection deploy form is dependent on the `snapshotFee` params, to compute max `proposerCut` value
- FE does not have the address of the space collection contract to retrieve that value

## 💊 Fixes / Solution

- Use trusted backed to return the `snapshotFee` value, from the smart contract
- Value should be cached, as it will very rarely change

Fix #95 

## 🚧 Changes

- Add GET `/api/nft-claimer` endpoint
- Endpoint should return a JSON response `{snapshotFee: X}`, with X being the `snapshotFee` value set in the spaceFactory smart contract
- Value is cached, and cache will refresh if created more than 1 hour ago

## 🛠️ Tests

- `curl localhost:3005/api/nft-claimer`
- It should return a JSON response with the snapshotFee

## Considerations

There is no error handling in case of rpc node failing for now. Waiting for some errors to know what kind of errors it will trigger, and how to handle them.

### CODE REVIEW

**Review by ChaituVR — APPROVED** [review]
utAck

File: src/lib/nftClaimer/utils.ts [hunk 1]
```diff
@@ -132,3 +137,43 @@ export function validateAddresses(addresses: Record<string, string>) {
...
+
+      if (result.lastUpdate > +new Date() - CACHE_TTL) {
+        return result.snapshotFee;
+      }
+    }
+
+    const providerOptions: Record<string, string> = {};
+    ['etherscan'].forEach(key => {
+      const envValue = process.env[`${key.toUpperCase()}_API_KEY`];
+      if (!!envValue) {
```
Comment (ChaituVR) [inline src/lib/nftClaimer/utils.ts:159]: why not just 
```suggestion
      if (envValue) {
```

File: src/lib/nftClaimer/utils.ts [hunk 2]
```diff
@@ -132,3 +137,43 @@ export function validateAddresses(addresses: Record<string, string>) {
...
+        providerOptions[key] = envValue;
+      }
+    });
+
+    const provider = getDefaultProvider(NFT_CLAIMER_NETWORK, providerOptions);
```
Comment (ChaituVR) [inline src/lib/nftClaimer/utils.ts:164]: We should use brovider here instead right?

**Review by ChaituVR — CHANGES_REQUESTED** [review]

File: src/lib/nftClaimer/utils.ts (hunk 2 above)
Comment (wa0x6e) [reply src/lib/nftClaimer/utils.ts:164]: Was not aware brovider can do that, will take a look

File: README.md [hunk 3]
```diff
@@ -144,6 +144,24 @@ Data are sourced from the json files with the same name, located in this repo `/
...
 Validate offchain data, and return a payload
 
+#### Get global data
+
+Retrieve global data from the smart contract (results are cached in the StorageEngine, with a 60min TTL).
+
+Send a `GET` request to `/api/nft-claimer`
+
+```bash
+curl -X POST localhost:3005/api/nft-claimer
```
Comment (ChaituVR) [inline README.md:154]:
```suggestion
curl -X GET localhost:3005/api/nft-claimer
```

File: src/lib/nftClaimer/utils.ts [hunk 4]
```diff
@@ -132,3 +137,43 @@ export function validateAddresses(addresses: Record<string, string>) {
...
+    if (typeof cache === 'string') {
+      const result = JSON.parse(cache);
+
+      if (result.lastUpdate > +new Date() - CACHE_TTL) {
+        return result.snapshotFee;
+      }
+    }
+
+    const providerOptions: Record<string, string> = {};
+    ['etherscan'].forEach(key => {
```
[the remark this answers is not in the document]
Comment (wa0x6e) [reply src/lib/nftClaimer/utils.ts]: removed, thanks to brovider

File: src/lib/nftClaimer/utils.ts [hunk 5]
```diff
@@ -132,3 +136,34 @@ export function validateAddresses(addresses: Record<string, string>) {
...
+        return result.snapshotFee;
+      }
+    }
+
+    const provider = snapshot.utils.getProvider(NFT_CLAIMER_NETWORK);
+    const contract = new Contract(
+      DEPLOY_CONTRACT,
+      ['function snapshotFee() public view returns (uint8)'],
+      provider
+    );
```
Comment (ChaituVR) [inline src/lib/nftClaimer/utils.ts:160]: Also we can move these out of this function as they will be always the same

Comment (wa0x6e) [reply src/lib/nftClaimer/utils.ts:146]: Cache removed

File: src/lib/nftClaimer/utils.ts [hunk 6]
```diff
@@ -132,3 +135,18 @@ export function validateAddresses(addresses: Record<string, string>) {
...
+
+export async function snapshotFee() {
+  try {
+    const provider = snapshot.utils.getProvider(NFT_CLAIMER_NETWORK);
+    const contract = new Contract(
+      DEPLOY_CONTRACT,
+      ['function snapshotFee() public view returns (uint8)'],
+      provider
+    );
+    return contract.snapshotFee();
```
Comment (ChaituVR) [inline src/lib/nftClaimer/utils.ts:147]: I mean like this, the provider and contract won't change unless we change env
```suggestion
const provider = snapshot.utils.getProvider(NFT_CLAIMER_NETWORK);
const contract = new Contract(
  DEPLOY_CONTRACT,
  ['function snapshotFee() public view returns (uint8)'],
  provider
);
export async function snapshotFee() {
  try {
    return contract.snapshotFee();
```

Comment (wa0x6e) [reply src/lib/nftClaimer/utils.ts:147]: Those variables are not used outside this function. Keeping them inside will save some memory, as they'll be instantiated only when needed.

Comment (ChaituVR) [reply src/lib/nftClaimer/utils.ts:147]: oh i see, thank you for explaining

**Review by ChaituVR — APPROVED** [review]
tAck

**Pull request merged**